backend/ecs_tasks/delete_files/parquet.py

```diff
@@ -15,10 +15,40 @@ def get_row_count(df):
     return len(df.index)
 
 
-def delete_from_dataframe(df, to_delete):
+def needs_flattening(to_delete):
+    """
+    Lookup the columns to identify if any are nested inside structs
+    (and therefore represented in a form like foo.bar).
+
+    Operating on flattened columns is necessary when operating with 
+    complex column types but it impacts performance during write as
+    we need to re-allocate the data in-memory as unflattened to 
+    preserve the initial schema's hierarchy after figuring out which rows
+    need deletion. This is why we check first: it allows to do the
+    re-allocation only if necessary.
+    """
+    return any(x['Column'].find(".") >= 0 for x in to_delete)
+
+def delete_from_table(table, to_delete, schema):
+    """
+    Deletes rows from a Arrow Table where any of the MatchIds is found as
+    value in any of the columns
+    """
+    needs_flattened_columns = needs_flattening(to_delete)
+    df = (table.flatten() if needs_flattened_columns else table).to_pandas()
+    initial_rows = get_row_count(df)
+    indexes_to_delete = []
     for column in to_delete:
-        df = df[~df[column["Column"]].isin(column["MatchIds"])]
-    return df
+        indexes = df[column["Column"]].isin(column["MatchIds"])
+        indexes_to_delete.append(indexes)
+        df = df[~indexes]
+    if needs_flattened_columns:
+        df = table.to_pandas()
+        for indexes in indexes_to_delete:
+            df=df[~indexes]
```

Comment:
For my benefit, could you explain what's happening in this block?

It looks like we're throwing away the value of `df` that's been built up during the previous loop then doing another similar loop. If that's the case it seems like there's room for optimisation?

But there could be some nuance I'm missing.

Reply:
Ok. I put a general comment in the `needs_flattening` method but I can put it back here referencing direct lines of code. The problem is that when you flatten the DF that gets converted from 
```
field1|field2|field3
foo|{x=1,y=2}|bar
```
to
```
field1|field2.x|field2.y|bar
foo|1|2|bar
```
and if I write as flattened, that won't be the same schema as when I read that. That's why I use the flattened for the find, but then I re-iterate writing on the same indexes.

In terms of the specific line `df = df[df~indexes]` on the flattened DF, the reason I do it is that if I have multiple columns, that would guarantee that while re-iterating as a part of the second for loop, I would delete the same indexes. Do you think we can optimise further there? I'm thinking maybe I can just merge indexes in the `indexes_to_delete` (instead of keeping it as array) and avoid executing the first `df = df[df~indexes]` command...

Reply:
Ah, ok. Thanks for the explanation.

I think this is fine as it is - I don't see obvious gains to trying to optimise it further.